// eslint-plugin-jest/lib/rules/no-export.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _util = require("./util");

var _default = {
  meta: {
    docs: {
      url: (0, _util.getDocsUrl)(__filename)
    },
    messages: {
      unexpectedExport: `Do not export from a test file.`
    },
    schema: []
  },

  create(context) {
    const exportNodes = [];
    let hasTestCase = false;
    return {
      'Program:exit'() {
        if (hasTestCase && exportNodes.length > 0) {
          for (var _i = 0, _exportNodes = exportNodes; _i < _exportNodes.length; _i++) {
            let node = _exportNodes[_i];
            context.report({
              node,
              messageId: 'unexpectedExport'
            });
          }
        }
      },

      CallExpression(node) {
        if ((0, _util.isTestCase)(node)) {
          hasTestCase = true;
        }
      },

      'ExportNamedDeclaration, ExportDefaultDeclaration'(node) {
        exportNodes.push(node);
      },

      'AssignmentExpression > MemberExpression'(node) {
        let object = node.object,
            property = node.property;

        if (object.type === 'MemberExpression') {
          var _object = object;
          object = _object.object;
          property = _object.property;
        }

        if (object.name === 'module' && /^exports?$/.test(property.name)) {
          exportNodes.push(node);
        }
      }

    };
  }

};
exports.default = _default;
