// eslint-plugin-jest/lib/rules/no-if.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _util = require("./util");

const isTestArrowFunction = node => node !== undefined && node.type === 'ArrowFunctionExpression' && (0, _util.isTestCase)(node.parent);

var _default = {
  meta: {
    docs: {
      description: 'Disallow conditional logic',
      category: 'Best Practices',
      recommended: false,
      uri: (0, _util.getDocsUrl)('jest/no-if')
    },
    messages: {
      noIf: 'Tests should not contain if statements.',
      noConditional: 'Tests should not contain conditional statements.'
    }
  },

  create(context) {
    const stack = [];

    function validate(node) {
      const lastElementInStack = stack[stack.length - 1];

      if (stack.length === 0 || lastElementInStack === false) {
        return;
      }

      const messageId = node.type === 'ConditionalExpression' ? 'noConditional' : 'noIf';
      context.report({
        messageId,
        node
      });
    }

    return {
      CallExpression(node) {
        stack.push((0, _util.isTestCase)(node));
      },

      FunctionExpression() {
        stack.push(false);
      },

      FunctionDeclaration() {
        stack.push(false);
      },

      ArrowFunctionExpression(node) {
        stack.push(isTestArrowFunction(node));
      },

      IfStatement: validate,
      ConditionalExpression: validate,

      'CallExpression:exit'() {
        stack.pop();
      },

      'FunctionExpression:exit'() {
        stack.pop();
      },

      'FunctionDeclaration:exit'() {
        stack.pop();
      },

      'ArrowFunctionExpression:exit'() {
        stack.pop();
      }

    };
  }

};
exports.default = _default;
