// eslint-plugin-jest/lib/rules/no-mocks-import.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _path = require("path");

var _util = require("./util");

const mocksDirName = '__mocks__';

const isMockPath = path => path.split(_path.posix.sep).includes(mocksDirName);

var _default = {
  meta: {
    docs: {
      url: (0, _util.getDocsUrl)(__filename)
    },
    messages: {
      noManualImport: `Mocks should not be manually imported from a ${mocksDirName} directory. Instead use jest.mock and import from the original module path.`
    },
    schema: []
  },

  create(context) {
    return {
      ImportDeclaration(node) {
        if (isMockPath(node.source.value)) {
          context.report({
            node,
            messageId: 'noManualImport'
          });
        }
      },

      'CallExpression[callee.name="require"]'(node) {
        if (node.arguments.length && node.arguments[0].value && isMockPath(node.arguments[0].value)) {
          context.report({
            loc: node.arguments[0].loc,
            messageId: 'noManualImport'
          });
        }
      }

    };
  }

};
exports.default = _default;
