// eslint-plugin-jest/lib/rules/no-truthy-falsy.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _util = require("./util");

var _default = {
  meta: {
    docs: {
      url: (0, _util.getDocsUrl)(__filename)
    },
    messages: {
      avoidMessage: 'Avoid {{methodName}}'
    },
    schema: []
  },

  create(context) {
    return {
      CallExpression(node) {
        if ((0, _util.expectCaseWithParent)(node) || (0, _util.expectNotCase)(node) || (0, _util.expectResolvesCase)(node) || (0, _util.expectRejectsCase)(node)) {
          const targetNode = node.parent.parent.type === 'MemberExpression' ? node.parent : node;
          const methodNode = (0, _util.method)(targetNode);
          const methodName = methodNode.name;

          if (methodName === 'toBeTruthy' || methodName === 'toBeFalsy') {
            context.report({
              data: {
                methodName
              },
              messageId: 'avoidMessage',
              node: methodNode
            });
          }
        }
      }

    };
  }

};
exports.default = _default;
