// eslint-plugin-jest/lib/rules/no-test-return-statement.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _util = require("./util");

const RETURN_STATEMENT = 'ReturnStatement';
const BLOCK_STATEMENT = 'BlockStatement';

const getBody = args => {
  if (args.length > 1 && (0, _util.isFunction)(args[1]) && args[1].body.type === BLOCK_STATEMENT) {
    return args[1].body.body;
  }

  return [];
};

var _default = {
  meta: {
    docs: {
      url: (0, _util.getDocsUrl)(__filename)
    },
    messages: {
      noReturnValue: 'Jest tests should not return a value.'
    },
    schema: []
  },

  create(context) {
    return {
      CallExpression(node) {
        if (!(0, _util.isTestCase)(node)) return;
        const body = getBody(node.arguments);
        const returnStmt = body.find(t => t.type === RETURN_STATEMENT);
        if (!returnStmt) return;
        context.report({
          messageId: 'noReturnValue',
          node: returnStmt
        });
      }

    };
  }

};
exports.default = _default;
